Sources/SwiftSyntax/SyntaxChildren.swift

```diff
@@ -10,305 +10,246 @@
-      self.parent = parent
-      self.previousChildInfo = endsOn
-    }
-
-    mutating func next() -> AbsoluteRawSyntax? {
-      while true {
-        let prevIdx = Int(previousChildInfo.indexInParent)
-        guard prevIdx > 0 else {
-          return nil
-        }
-        let child = parent.child(at: prevIdx-1)
-        previousChildInfo = previousChildInfo.reversedBySibling(child)
-        if let n = child, n.isPresent {
-          return AbsoluteRawSyntax(raw: n, info: previousChildInfo)
-        }
-      }
-    }
-  }
-
-  private let node: AbsoluteRawSyntax
-
-  init(_ absoluteRaw: AbsoluteRawSyntax) {
-    self.node = absoluteRaw
+    let endPosition = parent.info.advancedToEndOfChildren(parent.raw)
+    self.endIndex = SyntaxChildrenIndex(endPosition)
   }
 
   init(_ base: _SyntaxBase) {
     self.init(base.data.absoluteRaw)
   }
-
-  func makeIterator() -> Iterator {
-    return .init(parent: node)
-  }
-}
-
-/// Sequence of the sibling nodes following the provided node.
-struct PresentRawSyntaxNextSiblings: Sequence {
-  typealias Iterator = PresentRawSyntaxChildren.Iterator
-
-  private let node: _SyntaxBase
-
-  init(_ node: _SyntaxBase) {
-    self.node = node
-  }
-
-  func makeIterator() -> Iterator {
-    let rawParent = node.parent!.raw
-    let absoluteRaw = node.data.absoluteRaw
-    return .init(parent: rawParent,
-      startFrom: absoluteRaw.info.advancedBySibling(absoluteRaw.raw))
-  }
-}
-
-/// Sequence of the sibling nodes preceding the provided node, in reverse order.
-struct PresentRawSyntaxPreviousSiblings: Sequence {
-  typealias Iterator = ReversedPresentRawSyntaxChildren.Iterator
-
-  private let node: _SyntaxBase
-
-  init(_ node: _SyntaxBase) {
-    self.node = node
-  }
-
-  func makeIterator() -> Iterator {
-    let rawParent = node.parent!.raw
-    let absoluteRaw = node.data.absoluteRaw
-    return .init(parent: rawParent, endsOn: absoluteRaw.info)
-  }
 }
 
-/// Sequence of present children nodes of the provided `_SyntaxBase` node.
-struct SyntaxBaseChildren: Sequence {
-  struct Iterator: IteratorProtocol {
-    let parent: _SyntaxBase
-    var iterator: PresentRawSyntaxChildren.Iterator
-
-    init(node: _SyntaxBase) {
-      self.iterator = .init(parent: node.data.absoluteRaw)
-      self.parent = node
-    }
-
-    mutating func next() -> _SyntaxBase? {
-      guard let absoluteRaw = iterator.next() else { return nil }
-      let data = SyntaxData(absoluteRaw, parent: self.parent)
-      return makeSyntax(data)
+/// Collection that contains the `present` child nodes of an
+/// `AbsoluteRawSyntax` node.
+struct PresentRawSyntaxChildren: BidirectionalCollection {
+  typealias Element = AbsoluteRawSyntax
+  typealias Index = SyntaxChildrenIndex
+
+  /// The underlying collection which contains all children. The present
+  /// children are filtered from this collection.
+  private var allChildren: RawSyntaxChildren
+
+  let startIndex: SyntaxChildrenIndex
+  let endIndex: SyntaxChildrenIndex
```

Comment:
`startIndex` does some amount of traversing so probably ok to store as field but `endIndex` can be a calculated property.

Reply:
See comment above.

Reply:
Right now `PresentRawSyntaxChildren.endIndex` is initialized like this though:`self.endIndex = allChildren.endIndex`
It's not avoiding any calculation, it is storing the same value twice unnecessarily.